cpp/src/arrow/dataset/dataset.cc

```diff
@@ -31,21 +31,28 @@ namespace dataset {
 
 SimpleDataFragment::SimpleDataFragment(
     std::vector<std::shared_ptr<RecordBatch>> record_batches)
-    : record_batches_(std::move(record_batches)) {}
+    : SimpleDataFragment(std::move(record_batches), ScanOptions::Defaults()) {}
+
+SimpleDataFragment::SimpleDataFragment(
+    std::vector<std::shared_ptr<RecordBatch>> record_batches,
+    std::shared_ptr<ScanOptions> options)
+    : record_batches_(std::move(record_batches)), options_(std::move(options)) {}
 
 Status SimpleDataFragment::Scan(std::shared_ptr<ScanContext> scan_context,
                                 ScanTaskIterator* out) {
   // Make an explicit copy of record_batches_ to ensure Scan can be called
   // multiple times.
-  auto it = MakeVectorIterator(record_batches_);
+  auto batches_it = MakeVectorIterator(record_batches_);
 
   // RecordBatch -> ScanTask
-  auto fn = [](std::shared_ptr<RecordBatch> batch) -> std::unique_ptr<ScanTask> {
+  auto options = options_;
+  auto fn = [=](std::shared_ptr<RecordBatch> batch) -> std::unique_ptr<ScanTask> {
     std::vector<std::shared_ptr<RecordBatch>> batches{batch};
-    return ::arrow::internal::make_unique<SimpleScanTask>(std::move(batches));
+    return ::arrow::internal::make_unique<SimpleScanTask>(
+        std::move(batches), std::move(options), std::move(scan_context));
```

Comment:
As per comment above, you might want to copy instead of move (it's only a shared_ptr anyway). There's no test for it, and I'm not sure this behavior should be preserved.